src/algebra/direct_sum/decomposition.lean

```diff
@@ -0,0 +1,176 @@
+  inv_fun := direct_sum.coe_add_monoid_hom ℳ,
+  left_inv := decomposition.left_inv,
+  right_inv := decomposition.right_inv }
+
+@[simp] lemma decomposition.decompose'_eq : decomposition.decompose' = decompose ℳ := rfl
+
+@[simp] lemma decompose_symm_of {i : ι} (x : ℳ i) :
+  (decompose ℳ).symm (direct_sum.of _ i x) = x :=
+direct_sum.coe_add_monoid_hom_of ℳ _ _
+
+@[simp] lemma decompose_coe {i : ι} (x : ℳ i) :
+  decompose ℳ (x : M) = direct_sum.of _ i x :=
+by rw [←decompose_symm_of, equiv.apply_symm_apply]
+
+lemma decompose_of_mem {x : M} {i : ι} (hx : x ∈ ℳ i) :
+  decompose ℳ x = direct_sum.of (λ i, ℳ i) i ⟨x, hx⟩ :=
+decompose_coe _ ⟨x, hx⟩
+
+lemma decompose_of_mem_same {x : M} {i : ι} (hx : x ∈ ℳ i) :
+  (decompose ℳ x i : M) = x :=
+by rw [decompose_of_mem _ hx, direct_sum.of_eq_same, subtype.coe_mk]
+
+lemma decompose_of_mem_ne {x : M} {i j : ι} (hx : x ∈ ℳ i) (hij : i ≠ j):
+  (decompose ℳ x j : M) = 0 :=
+by rw [decompose_of_mem _ hx, direct_sum.of_eq_of_ne _ _ _ _ hij,
+  add_submonoid_class.coe_zero]
+
+/-- If `M` is graded by `ι` with degree `i` component `ℳ i`, then it is isomorphic as
+an additive monoid to a direct sum of components. -/
+@[simps {fully_applied := ff}]
+def decompose_add_equiv : M ≃+ ⨁ i, ℳ i := add_equiv.symm
+{ map_add' := map_add (direct_sum.coe_add_monoid_hom ℳ),
+  ..(decompose ℳ).symm }
+
+@[simp] lemma decompose_zero : decompose ℳ (0 : M) = 0 := map_zero (decompose_add_equiv ℳ)
+@[simp] lemma decompose_symm_zero : (decompose ℳ).symm 0 = (0 : M) :=
+map_zero (decompose_add_equiv ℳ).symm
+
+@[simp] lemma decompose_add (x y : M) : decompose ℳ (x + y) = decompose ℳ x + decompose ℳ y :=
+map_add (decompose_add_equiv ℳ) x y
+@[simp] lemma decompose_symm_add (x y : ⨁ i, ℳ i) :
+  (decompose ℳ).symm (x + y) = (decompose ℳ).symm x + (decompose ℳ).symm y :=
+map_add (decompose_add_equiv ℳ).symm x y
+
+@[simp] lemma decompose_sum {ι'} (s : finset ι') (f : ι' → M) :
+  decompose ℳ (∑ i in s, f i) = ∑ i in s, decompose ℳ (f i) :=
+map_sum (decompose_add_equiv ℳ) f s
+@[simp] lemma decompose_symm_sum {ι'} (s : finset ι') (f : ι' → ⨁ i, ℳ i) :
+  (decompose ℳ).symm (∑ i in s, f i) = ∑ i in s, (decompose ℳ).symm (f i) :=
+map_sum (decompose_add_equiv ℳ).symm f s
+
+/-- This instance produces the instance needed by `dfinsupp.support` -/
+-- instance set_like_decidable_ne_zero [Π x : M, decidable (x ≠ 0)] :
+--   Π i (x : ℳ i), decidable (x ≠ 0) :=
+-- λ i x, decidable_of_iff _ add_submonoid_class.coe_eq_zero.not
+
+lemma sum_support_decompose [Π i (x : ℳ i), decidable (x ≠ 0)] (r : M) :
+  ∑ i in (decompose ℳ r).support, (decompose ℳ r i : M) = r :=
+begin
+  conv_rhs { rw [←(decompose ℳ).symm_apply_apply r,
+    ←sum_support_of (λ i, (ℳ i)) (decompose ℳ r)] },
+  rw [decompose_symm_sum],
+  simp_rw decompose_symm_of,
+end
+
+end add_comm_monoid
+
+-- the `-` in the statements below doesn't resolve without this line
+instance add_comm_group_set_like [add_comm_group M] [set_like σ M] [add_subgroup_class σ M]
```

Comment:
Is this a known issue (the same one as the missing finsupp instances)? Or could you add a bit more explanation why it breaks?

Reply:
What missing finsupp instances are you referring to?

Reply:
I added an extended comment here; it turns out there is a way to make this work, but it requires using `@eq` to force lean to get the type right.